File: hw/vizcaino/u3/hw25CompositePattern/hw25CompositePatternPython/supervisor.py
from employee import Employee

class Supervisor(Employee):
    def __init__(self):
        super().__init__()
        self.direct_reports = []

    def state_name(self):
        super().state_name()
        if len(self.direct_reports) > 0:
            for report in self.direct_reports:
                report.state_name()

    def add(self, employee: Employee):
        self.direct_reports.append(employee)

File: hw/vizcaino/u3/hw25CompositePattern/hw25CompositePatternJavaScript/supervisor.js
const Employee = require('./employee');

class Supervisor extends Employee {
    constructor() {
        super();
        if (this.constructor === Supervisor) {
            throw new Error("Cannot instantiate abstract class Supervisor");
        }
        this.directReports = [];
    }

    stateName() {
        super.stateName();
        if (this.directReports.length > 0) {
            for (let i = 0; i < this.directReports.length; i++) {
                this.directReports[i].stateName();
            }
        }
    }

    add(employee) {
        this.directReports.push(employee);
    }
}

module.exports = Supervisor;